### ISSUE TITLE
test failure

### ISSUE THREAD

**Issue by migueldvb** [RobertLuptonTheGood/eups#55]
I am getting a few test failures on master with python 2.7 in the testAll.py suite. The output
of the `python testAll.py` command is:

```
..E......F...EE.....F..E..F.................E.EFF.E..E......................................................
======================================================================
ERROR: testDeclare (testCmd.CmdTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testCmd.py", line 251, in testDeclare
    self.assertEqual(cmd.run(), 0)
  File "/usr/local/share/eups/python/eups/cmd.py", line 240, in run
    return self.execute()
  File "/usr/local/share/eups/python/eups/cmd.py", line 167, in execute
    return ecmd.run()
  File "/usr/local/share/eups/python/eups/cmd.py", line 240, in run
    return self.execute()
  File "/usr/local/share/eups/python/eups/cmd.py", line 1324, in execute
    tag=self.opts.tag, eupsenv=myeups)
  File "/usr/local/share/eups/python/eups/app.py", line 490, in declare
    tablefile, externalFileList=externalFileList, tag=tag)
  File "/usr/local/share/eups/python/eups/Eups.py", line 2626, in declare
    self._databaseFor(eupsPathDir, dbpath).declare(product)
  File "/usr/local/share/eups/python/eups/db/Database.py", line 468, in declare
    versionFile.write(trimDir)
  File "/usr/local/share/eups/python/eups/db/VersionFile.py", line 486, in write
    if trimDir and eups.utils.commonpath([trimDir, value]) == trimDir:
  File "/usr/local/share/eups/python/eups/utils.py", line 865, in commonpath
    raise ValueError("Can't mix absolute and relative paths")
ValueError: Can't mix absolute and relative paths

======================================================================
ERROR: testRemove (testCmd.CmdTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testCmd.py", line 330, in testRemove
    eups.Eups().declare("newprod", "2.0", pdir20)
  File "/usr/local/share/eups/python/eups/Eups.py", line 2626, in declare
    self._databaseFor(eupsPathDir, dbpath).declare(product)
  File "/usr/local/share/eups/python/eups/db/Database.py", line 468, in declare
    versionFile.write(trimDir)
  File "/usr/local/share/eups/python/eups/db/VersionFile.py", line 486, in write
    if trimDir and eups.utils.commonpath([trimDir, value]) == trimDir:
  File "/usr/local/share/eups/python/eups/utils.py", line 865, in commonpath
    raise ValueError("Can't mix absolute and relative paths")
ValueError: Can't mix absolute and relative paths

======================================================================
ERROR: testUses (testCmd.CmdTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testCmd.py", line 218, in testUses
    self.assertEqual(cmd.run(), 0)
  File "/usr/local/share/eups/python/eups/cmd.py", line 240, in run
    return self.execute()
  File "/usr/local/share/eups/python/eups/cmd.py", line 167, in execute
    return ecmd.run()
  File "/usr/local/share/eups/python/eups/cmd.py", line 240, in run
    return self.execute()
  File "/usr/local/share/eups/python/eups/cmd.py", line 846, in execute
    tags=self.opts.tag, pickleFile=self.opts.pickleFile)
  File "/usr/local/share/eups/python/eups/app.py", line 307, in printUses
    eupsenv.setPreferredTags(tags)
  File "/usr/local/share/eups/python/eups/Eups.py", line 664, in setPreferredTags
    self._kindlySetPreferredTags(tags, True)
  File "/usr/local/share/eups/python/eups/Eups.py", line 699, in _kindlySetPreferredTags
    raise TagNotRecognized(str(notokay), msg="Unsupported tag(s): " + ", ".join(notokay))
TagNotRecognized: Unsupported tag(s): newest

======================================================================
ERROR: testDeclare (testDb.DatabaseTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testDb.py", line 753, in testDeclare
    self.db.declare(base2)
  File "/usr/local/share/eups/python/eups/db/Database.py", line 468, in declare
    versionFile.write(trimDir)
  File "/usr/local/share/eups/python/eups/db/VersionFile.py", line 486, in write
    if trimDir and eups.utils.commonpath([trimDir, value]) == trimDir:
  File "/usr/local/share/eups/python/eups/utils.py", line 865, in commonpath
    raise ValueError("Can't mix absolute and relative paths")
ValueError: Can't mix absolute and relative paths

======================================================================
ERROR: testDeclare (testEups.EupsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testEups.py", line 253, in testDeclare
    self.eups.declare("newprod", "1.0", pdir10, testEupsStack, table)
  File "/usr/local/share/eups/python/eups/Eups.py", line 2626, in declare
    self._databaseFor(eupsPathDir, dbpath).declare(product)
  File "/usr/local/share/eups/python/eups/db/Database.py", line 468, in declare
    versionFile.write(trimDir)
  File "/usr/local/share/eups/python/eups/db/VersionFile.py", line 486, in write
    if trimDir and eups.utils.commonpath([trimDir, value]) == trimDir:
  File "/usr/local/share/eups/python/eups/utils.py", line 865, in commonpath
    raise ValueError("Can't mix absolute and relative paths")
ValueError: Can't mix absolute and relative paths

======================================================================
ERROR: testFindProduct (testEups.EupsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testEups.py", line 158, in testFindProduct
    tag = self.eups.tags.getTag("newest")
  File "/usr/local/share/eups/python/eups/tags.py", line 123, in getTag
    raise TagNotRecognized(str(tag))
TagNotRecognized: Tag not recognized: newest

======================================================================
ERROR: testRemove (testEups.EupsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testEups.py", line 491, in testRemove
    self.eups.declare("newprod", "2.0", pdir20)
  File "/usr/local/share/eups/python/eups/Eups.py", line 2626, in declare
    self._databaseFor(eupsPathDir, dbpath).declare(product)
  File "/usr/local/share/eups/python/eups/db/Database.py", line 468, in declare
    versionFile.write(trimDir)
  File "/usr/local/share/eups/python/eups/db/VersionFile.py", line 486, in write
    if trimDir and eups.utils.commonpath([trimDir, value]) == trimDir:
  File "/usr/local/share/eups/python/eups/utils.py", line 865, in commonpath
    raise ValueError("Can't mix absolute and relative paths")
ValueError: Can't mix absolute and relative paths

======================================================================
ERROR: testDetectOutOfSync (testEups.EupsCacheTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testEups.py", line 549, in testDetectOutOfSync
    e1.declare("newprod", "1.0", pdir10, testEupsStack, table)
  File "/usr/local/share/eups/python/eups/Eups.py", line 2626, in declare
    self._databaseFor(eupsPathDir, dbpath).declare(product)
  File "/usr/local/share/eups/python/eups/db/Database.py", line 468, in declare
    versionFile.write(trimDir)
  File "/usr/local/share/eups/python/eups/db/VersionFile.py", line 486, in write
    if trimDir and eups.utils.commonpath([trimDir, value]) == trimDir:
  File "/usr/local/share/eups/python/eups/utils.py", line 865, in commonpath
    raise ValueError("Can't mix absolute and relative paths")
ValueError: Can't mix absolute and relative paths

======================================================================
FAIL: testList (testCmd.CmdTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testCmd.py", line 147, in testList
    self.assertEquals(self.out.getvalue(), outall)
AssertionError: 'base                  1.0        \tcurrent\nbase                  2.0        \ncfitsio               3006.2     \tcurrent\ndoxygen               1.5.7.1    \tcurrent\neigen                 2.0.0      \tcurrent\nmpich2                1.0.5p4    \tcurrent\npython                2.5.2      \tcurrent\npython                2.6        \ntcltk                 8.5a4      \tcurrent' != 'cfitsio               3006.2     \tcurrent\ndoxygen               1.5.7.1    \tcurrent\neigen                 2.0.0      \tcurrent\nmpich2                1.0.5p4    \tcurrent\npython                2.5.2      \tcurrent\npython                2.6        \ntcltk                 8.5a4      \tcurrent'

======================================================================
FAIL: testSetVersion (testDb.ChainFileTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testDb.py", line 405, in testSetVersion
    self.assert_(bool(flavor['declarer']))
AssertionError: False is not true

======================================================================
FAIL: testFindProductNames (testDb.DatabaseTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testDb.py", line 500, in testFindProductNames
    self.assertEquals(len(prods), 6)
AssertionError: 7 != 6

======================================================================
FAIL: testInit (testEups.EupsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testEups.py", line 83, in testInit
    self.assert_(tag in tags)
AssertionError: False is not true

======================================================================
FAIL: testList (testEups.EupsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mdevalbo/project/LSST/eups/tests/testEups.py", line 404, in testList
    self.assertEquals(prods[0].version, "2.5.2")
AssertionError: '2.6' != '2.5.2'

----------------------------------------------------------------------
Ran 108 tests in 4.329s

FAILED (failures=5, errors=8)
```

**Comment by timj** [comment]
I get some of those failures with master, but not all of them.
```
.........E....E.............................F.EFF...........................................................
======================================================================
ERROR: testList (testCmd.CmdTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/timj/work/lsst/eups/tests/testCmd.py", line 163, in testList
    self.assertEqual(cmd.run(), 0)
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/cmd.py", line 240, in run
    return self.execute()
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/cmd.py", line 167, in execute
    return ecmd.run()
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/cmd.py", line 240, in run
    return self.execute()
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/cmd.py", line 509, in execute
    raw=self.opts.raw
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/app.py", line 54, in printProducts
    checkTagsList(eupsenv, tags)
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/tags.py", line 545, in checkTagsList
    msg="Unsupported tag(s): %s" % ", ".join(map(lambda t: str(t), badtags)))
TagNotRecognized: Unsupported tag(s): newest

======================================================================
ERROR: testUses (testCmd.CmdTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/timj/work/lsst/eups/tests/testCmd.py", line 218, in testUses
    self.assertEqual(cmd.run(), 0)
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/cmd.py", line 240, in run
    return self.execute()
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/cmd.py", line 167, in execute
    return ecmd.run()
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/cmd.py", line 240, in run
    return self.execute()
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/cmd.py", line 846, in execute
    tags=self.opts.tag, pickleFile=self.opts.pickleFile)
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/app.py", line 307, in printUses
    eupsenv.setPreferredTags(tags)
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/Eups.py", line 664, in setPreferredTags
    self._kindlySetPreferredTags(tags, True)
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/Eups.py", line 699, in _kindlySetPreferredTags
    raise TagNotRecognized(str(notokay), msg="Unsupported tag(s): " + ", ".join(notokay))
TagNotRecognized: Unsupported tag(s): newest

======================================================================
ERROR: testFindProduct (testEups.EupsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/timj/work/lsst/eups/tests/testEups.py", line 159, in testFindProduct
    tag = self.eups.tags.getTag("newest")
  File "/Users/timj/work/lsstsw/eups/1.5.9/python/eups/tags.py", line 123, in getTag
    raise TagNotRecognized(str(tag))
TagNotRecognized: Tag not recognized: newest

======================================================================
FAIL: testDeclare (testEups.EupsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/timj/work/lsst/eups/tests/testEups.py", line 301, in testDeclare
    self.assertEquals(self.eups.findProduct("newprod", self.eups.tags.getTag("beta")).dir, pdir10)
AssertionError: '/Users/timj/work/lsst/eups/tests/Linux/newprod/1.1' != '/Users/timj/work/lsst/eups/tests/Linux/newprod/1.0'

======================================================================
FAIL: testInit (testEups.EupsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/timj/work/lsst/eups/tests/testEups.py", line 84, in testInit
    self.assert_(tag in tags)
AssertionError: False is not true

======================================================================
FAIL: testList (testEups.EupsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/timj/work/lsst/eups/tests/testEups.py", line 405, in testList
    self.assertEquals(prods[0].version, "2.5.2")
AssertionError: '2.6' != '2.5.2'

----------------------------------------------------------------------
Ran 108 tests in 2.785s

FAILED (failures=3, errors=3)
```

**Comment by timj** [comment]
I must be doing something wrong because I can't get tests to pass even if I check out older releases. I've gone back to 1.5.5. @RobertLuptonTheGood @mjuric I can't see any instructions for how to build and run tests so I'm doing the following:
```
% ./configure --prefix="`pwd`/tmp/eups/" --with-python=`which python` --with-eups="`pwd`/tmp/stack"
% make
% make test
```

**Comment by migueldvb** [comment]
@timj thank you for running the tests, are you using python 2.7? It looks like the python interpreter used in the Makefile to run the tests should be set as EUPS_PYTHON instead of the default python.

**Comment by timj** [comment]
I'm using python 2.7.

I think my test results differed because my `PYTHONPATH` had my installed EUPS so regardless of my checkouts I was always using the same EUPS.

The `commonpath` error was probably introduced by commit 2d484e3638a5a4d6065a1f42f38689e247ff8ae6

**Comment by mjuric** [comment]
Hi @timj @migueldvb -- I'm not sure if the test suite is up-to-date. @RobertLuptonTheGood ?

**Comment by timj** [comment]
Hmmm. I was not brave enough to release a 1.5.10 or merge in #52 without knowing that the tests still worked :smile:

**Comment by RobertLuptonTheGood** [comment]
On my Christmas Project List

**Comment by timj** [comment]
@mjuric I don't really understand why this ticket was closed? The EUPS test suite still fails for me in the same way as described in this ticket.

**Comment by timj** [comment]
It's not just me and the tests also fail for @r-owen so I'm reopening.

**Comment by mjuric** [comment]
I (mis?)understood this issue as "Fix the test suite so that it fails the same on Py3 branch as it did before those changes". I'm OK with keeping it opened.

**Comment by r-owen** [comment]
I think the ticket was intended to be "please fix the tests". I hope so!

**Comment by timj** [comment]
Also my assumption. These tests were broken before we even considered adopting Python3. They've been broken for a while now and it causes much pain when trying to work out if a new fix is really working.

**Comment by r-owen** [comment]
#75 may be relevant to the "ValueError: Can't mix absolute and relative paths" failures.

It would also be nice if one could run the tests in some simple way after doing "make". I realize eups somehow has to be setup, but I hope it can be the one in the directory where it just got built, rather than requiring it be installed elsewhere and then setting up the installed version.

**Comment by mjuric** [comment]
It's possible, but I made sure that test failures before/after the Python 3 merge were the same.

It is possible, though, that the test was uncovering the issue on Python 2, but the actual code somehow still managed to work, while with Py3 changes the code started triggering the problem as well.

**Comment by r-owen** [comment]
I fixed some errors and cleaned up test code on branch "fix-tests". The results are:
Ran 111 tests in 4.014s
FAILED (failures=1, errors=5)

All but one of these errors seems to be the same issue: the "newest" tag is not recognized (though once that problem is fixed more errors may be exposed, since tests stop on first failure). The remaining error is as follows:

    ======================================================================
    ERROR: testDeclare (testEups.EupsTestCase)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/Users/rowen/UW/LSST/eups_git/tests/testEups.py", line 369, in testDeclare
        self.eups.undeclare("newprod")
      File "/Users/rowen/UW/LSST/lsstsw/eups/git/python/eups/Eups.py", line 2742, in undeclare
        (productName, "\" \"".join(versionList)))
    EupsException: Product newprod has versions "1.0" "tag:beta"; please choose one and try again

I think this reflects a real problem. I have noticed for some time now that I sometimes end up with a package declared with a version named "tag:<something>" when I was trying to tag an existing version as "<something>" (usually my username). This error appears to be showing the same behavior.

**Comment by r-owen** [comment]
I have a few questions about the tests:

- Why does testCommon.py set environ["EUPS_DIR"]? It doesn't seem to affect the tests, and the value it sets (*pkgroot*`/tests`) does not contain an executable eups. This looks like a mistake and I think the line should simply be removed.
- Would it make sense for "make test" to be able to run the tests without the user explicitly running the eups setup script? For instance if EUPS_DIR was not set it could be set to *pkgroot*/bin but it would be safer to run the appropriate setup script, which sounds a bit messy.

**Comment by r-owen** [comment]
Pull request #78 fixes all but one test

**Comment by mjuric** [comment]
What I think is a fix for the `declare` issue (revealed by the unit test failure) is on #81 .

**Comment by mjuric** [comment]
@migueldvb, I think I know where the test failures you're seeing are coming from; I have fixes for all of them, will likely push it later today (need to break up the patches first). One of the problems is the changed behavior of `isSubpath` -- there's code in `VersionFile` that assumes it doesn't do any `realpath` transformations, and does something like:
```
if isSubpath(a, b):
   relpath = b[len(a)]
```
I'll implement a fix for that later today.

**Comment by r-owen** [comment]
Thank you Mario. I assume you already know about it, but `os.path` has a `relpath` function that might be just the thing here.

**Comment by mjuric** [comment]
I merged the `eups declare` fix in #81.

Assuming @migueldvb reports no problems with #82, we may be finally able to close this issue :).

**Comment by mjuric** [comment]
Resolved by #82. If new issues arise, open a separate issue.
